vowpalwabbit/core/src/reductions/bfgs.cc

```diff
@@ -1111,66 +1111,91 @@ void save_load(bfgs& b, io_buf& model_file, bool read, bool text)
-
   bool bfgs_option = false;
   int local_m = 0;
   float local_rel_threshold = 0.f;
   bool local_hessian_on = false;
+  uint64_t early_terminate = 0;
+  bool bfgs_enabled = false;
+};
+
+std::unique_ptr<options_bfgs_v1> get_bfgs_options_instance(
+    const VW::workspace& all, VW::io::logger&, options_i& options)
+{
+  auto bfgs_opts = VW::make_unique<options_bfgs_v1>();
+  option_group_definition conjugate_gradient_options("[Reduction] Conjugate Gradient");
+  conjugate_gradient_options.add(make_option("conjugate_gradient", bfgs_opts->conjugate_gradient)
+                                     .keep()
+                                     .necessary()
+                                     .help("Use conjugate gradient based optimization"));
   option_group_definition bfgs_options("[Reduction] LBFGS and Conjugate Gradient");
-  bfgs_options.add(
-      make_option("bfgs", bfgs_option).keep().necessary().help("Use conjugate gradient based optimization"));
-  bfgs_options.add(make_option("hessian_on", local_hessian_on).help("Use second derivative in line search"));
-  bfgs_options.add(make_option("mem", local_m).default_value(15).help("Memory in bfgs"));
-  bfgs_options.add(make_option("termination", local_rel_threshold).default_value(0.001f).help("Termination threshold"));
+  bfgs_options
+      .add(make_option("bfgs", bfgs_opts->bfgs_option)
+               .keep()
+               .necessary()
+               .help("Use conjugate gradient based optimization"))
+      .add(make_option("hessian_on", bfgs_opts->local_hessian_on).help("Use second derivative in line search"))
+      .add(make_option("mem", bfgs_opts->local_m).default_value(15).help("Memory in bfgs"))
+      .add(make_option("termination", bfgs_opts->local_rel_threshold)
+               .default_value(0.001f)
+               .help("Termination threshold"));
 
   auto conjugate_gradient_enabled = options.add_parse_and_check_necessary(conjugate_gradient_options);
-  auto bfgs_enabled = options.add_parse_and_check_necessary(bfgs_options);
-  if (!conjugate_gradient_enabled && !bfgs_enabled) { return nullptr; }
-  if (conjugate_gradient_enabled && bfgs_enabled) { THROW("'conjugate_gradient' and 'bfgs' cannot be used together."); }
-
-  b->all = &all;
-  b->wolfe1_bound = 0.01;
-  b->first_hessian_on = true;
-  b->first_pass = true;
-  b->gradient_pass = true;
-  b->preconditioner_pass = true;
-  b->backstep_on = false;
-  b->final_pass = all.numpasses;
-  b->no_win_counter = 0;
-
-  if (bfgs_enabled)
+  bfgs_opts->bfgs_enabled = options.add_parse_and_check_necessary(bfgs_options);
+  if (!conjugate_gradient_enabled && !bfgs_opts->bfgs_enabled) { return nullptr; }
+  if (conjugate_gradient_enabled && bfgs_opts->bfgs_enabled)
+  {
+    THROW("'conjugate_gradient' and 'bfgs' cannot be used together.");
+  }
+  if (!all.holdout_set_off)
+  {
+    bfgs_opts->early_terminate = options.get_typed_option<uint64_t>("early_terminate").value();
```

Comment:
we want to try and minimize these "digging out the value" operations we need to do. If a reduction has a dependency on an option it is best to put it in its option group

Reply:
I agree these should be in the option group